tests/pip_deps/pip_deps.bzl

```diff
@@ -0,0 +1,55 @@
+""" A demo implementation for pip_deps which provides @unpinned_pip//:pin. """
+
+load("@rules_python//python:pip.bzl", "pip_parse")
+
+def _requirements_in_impl(repository_ctx):
+    repository_ctx.file(
+        "requirements.in",
+        content = "".join(["{package} == {version}\n".format(
+            package = package,
+            version = version,
+        ) for (package, version) in repository_ctx.attr.packages.items()]),
+    )
+    repository_ctx.file("WORKSPACE", content = "")
+    repository_ctx.file("BUILD", content = """
+load("@rules_python//python:pip.bzl", "compile_pip_requirements")
+
+compile_pip_requirements(
+    name = "pin",
+    extra_args = ["--allow-unsafe"],
+    requirements_in = "requirements.in",
+    requirements_txt = "@{workspace_name}//{package}:{name}",
+)
+""".format(
+        workspace_name = repository_ctx.attr.requirements_lock.workspace_name,
+        package = repository_ctx.attr.requirements_lock.package,
+        name = repository_ctx.attr.requirements_lock.name,
+    ))
+
+_requirements_in = repository_rule(
+    implementation = _requirements_in_impl,
+    attrs = {
+        "packages": attr.string_dict(),
+        "requirements_lock": attr.label(allow_single_file = True),
+    },
+)
+
+def pip_deps(
+        *,
+        name = "pip",
+        packages = {},
+        requirements_lock_target = Label("//:requirements_txt"),
+        requirements_lock_file = Label("//:requirements.txt"),
+        python_interpreter_target = None,
+        **kwargs):
+    _requirements_in(
+        name = "unpinned_" + name,
```

Comment:
I think the term "unpinned" may be inaccurate here. I'm not sure how it compares with rules_jvm_external, but here, the existing locked requirements file is still an input to pip-compile, so any already-pinned transitive dependencies will remain pinned to the same version. Only versions which no longer satisfy the constraints in requirements.in will actually change.

If the behavior is different from rules_jvm_external, then maybe another name would be better?

Reply:
> I think the term "unpinned" may be inaccurate here. I'm not sure how it compares with rules_jvm_external, but here, the existing locked requirements file is still an input to pip-compile, so any already-pinned transitive dependencies will remain pinned to the same version. Only versions which no longer satisfy the constraints in requirements.in will actually change.
> 

The input to _requirements_in is `packages` and output is a `requirements.in` file. None of them have anything to do with the locked requirements file.

> If the behavior is different from rules_jvm_external, then maybe another name would be better?

Reply:
There's a requirements_lock attribute here that gets passed to pip_compile. Am I missing something?